From 61fa742ca33177773a867dc0e99018cec51f5c77 Mon Sep 17 00:00:00 2001
From: LIU Hao <lh_mouse@126.com>
Date: Mon, 9 Feb 2026 13:18:21 +0800
Subject: [PATCH] Revert "Restore old behaviour of windres so that options
 containing spaces are not enclosed in double quotes."

This reverts commit 749c700282097cf679ff019a9674d7c762f48619.
---
 binutils/doc/binutils.texi |  7 ++++---
 binutils/windres.c         | 26 +++++++++++++++++++++++++-
 2 files changed, 29 insertions(+), 4 deletions(-)

diff --git a/binutils/doc/binutils.texi b/binutils/doc/binutils.texi
index e52ec69616a..e26b22f033e 100644
--- a/binutils/doc/binutils.texi
+++ b/binutils/doc/binutils.texi
@@ -4566,7 +4566,8 @@ format, which is the first one listed by the @option{--help} option.
 @item --preprocessor @var{program}
 When @command{windres} reads an @code{rc} file, it runs it through the C
 preprocessor first.  This option may be used to specify the preprocessor
-to use.  The default preprocessor is @code{gcc}.
+to use, including any leading arguments.  The default preprocessor
+argument is @code{gcc}.
 
 @item --preprocessor-arg @var{option}
 When @command{windres} reads an @code{rc} file, it runs it through
@@ -4577,8 +4578,8 @@ preprocessor command line.
 If the @option{--preprocessor} option has not been specified then a
 default set of preprocessor arguments will be used, with any
 @option{--preprocessor-arg} options being placed after them on the
-command line.  These default arguments are @code{-E},
-@code{-xc-header} and @code{-DRC_INVOKED}.
+command line.  These default arguments are @code{-E -xc-header
+-DRC_INVOKED}.
 
 @item -I @var{directory}
 @itemx --include-dir @var{directory}
diff --git a/binutils/windres.c b/binutils/windres.c
index 11cf3d5652c..c4fe83df078 100644
--- a/binutils/windres.c
+++ b/binutils/windres.c
@@ -690,13 +690,37 @@ quot (const char *string)
       buf = (char *) xmalloc (buflen);
     }
 
+#if defined (_WIN32) && !defined (__CYGWIN__)
+  /* For Windows shells, quote "like this".   */
+  {
+    bool quoted = false;
+
+    dest = buf;
+    if (strchr (string, ' '))
+      {
+	quoted = true;
+	*dest++ = '"';
+      }
+
+    for (src = string; *src; src++, dest++)
+      {
+	/* Escape-protect embedded double quotes.  */
+	if (quoted && *src == '"')
+	  *dest++ = '\\';
+	*dest = *src;
+      }
+
+    if (quoted)
+      *dest++ = '"';
+  }
+#else
   for (src = string, dest = buf; *src; src++, dest++)
     {
       if (*src == '(' || *src == ')' || *src == ' ' || *src == '<' || *src == '>')
 	*dest++ = '\\';
       *dest = *src;
     }
-
+#endif
   *dest = 0;
   return buf;
 }
-- 
2.53.0

